File: 5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */
class Solution {
public:
    string longestPalindrome(string s) {
        if(s == "")
            return "";
        int maxLength = 1;
        string ret = s.substr(0,1);
        for(int i = 1;i < s.size(); i++){
            int j = 1;
            for(j = 1;;j++){
                if(i-j < 0 || i+j >= s.size() || s[i-j] != s[i+j]){
                    j--;
                    break;
                }
            }
            if(maxLength < 1+2*j){
                maxLength = 1+2*j;
                ret = s.substr(i-j, maxLength);
            }
            maxLength = (maxLength<1+2*j)?1+2*j:maxLength;
            if(s[i] == s[i-1]){
                for(j = 1;;j++){
                    if(i-j-1 < 0 || i+j >= s.size() || s[i-j-1] != s[i+j]){
                        j--;
                        break;
                    }
                }
                if(maxLength < 2+2*j){
                    maxLength = 2+2*j;
                    ret = s.substr(i-j-1, maxLength);
                }
            }
        }
        return ret;
    }
};

File: 5.最长回文子串.rs
/*
 * @lc app=leetcode.cn id=5 lang=rust
 *
 * [5] 最长回文子串
 */

// @lc code=start
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.is_empty() {
            return "".to_string();
        }
        let mut f = [[0;1001];1001];
    
        let bytes = s.as_bytes();
        let mut result = String::new();
        for l in 0..s.len() {
            for start in 0..s.len() {
                if l == 0 {
                    f[start][start + l] = 1;
                    if f[start][start + l] > result.len() {
                        result = s[start..start + l + 1].to_string();
                    }
                } else if start + l < bytes.len() && bytes[start] == bytes[start + l] {
                    if start + 1 <= start + l - 1 && f[start + 1][start + l - 1] > 0 {
                        f[start][start + l] = f[start + 1][start + l - 1] + 2;
                        if f[start][start + l] > result.len() {
                            result = s[start..start + l + 1].to_string();
                        }
                    } else if start + 1 > start + l - 1 {
                        f[start][start + l] = 2;
                        if f[start][start + l] > result.len() {
                            result = s[start..start + l + 1].to_string();
                        }
                    }
                }
            }
        }
    
        return result;
    }
}
// @lc code=end

